docs/metrics.md

```diff
@@ -0,0 +1,561 @@
+| [Asynchronous UpDownCounter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-updowncounter) | [`ObservableUpDownCounter`](https://docs.rs/opentelemetry/latest/opentelemetry/metrics/struct.ObservableUpDownCounter.html) |
+| [Counter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#counter) | [`Counter`](https://docs.rs/opentelemetry/latest/opentelemetry/metrics/struct.Counter.html) |
+| [Gauge](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#gauge) | [`Gauge`](https://docs.rs/opentelemetry/latest/opentelemetry/metrics/struct.Gauge.html) |
+| [Histogram](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#histogram) | [`Histogram`](https://docs.rs/opentelemetry/latest/opentelemetry/metrics/struct.Histogram.html) |
+| [UpDownCounter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#updowncounter) | [`UpDownCounter`](https://docs.rs/opentelemetry/latest/opentelemetry/metrics/struct.UpDownCounter.html) |
+
+:stop_sign: You should avoid creating instruments (e.g. `Counter`) too
+frequently. Instruments are fairly expensive and meant to be reused. For most
+applications, instruments can be created once and re-used. Instruments can also
+be cloned to create multiple handles to the same instrument, but the cloning
+should not be on hot path, but instead the cloned instance should be stored and
+re-used.
+
+:stop_sign: You should avoid invalid instrument names.
+
+> [!NOTE] OpenTelemetry will not collect metrics from instruments that are using
+> invalid names. Refer to the [OpenTelemetry
+  Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-name-syntax)
+  for the valid syntax.
+
+:stop_sign: You should avoid changing the order of attributes while reporting
+measurements.
+
+> [!WARNING] The last line of code has bad performance since the attributes are
+> not following the same order as before:
+
+```rust
+let counter = meter.u64_counter("example_counter").build();
+counter.add(2, &[KeyValue::new("color", "red"), KeyValue::new("name", "apple")]);
+counter.add(3, &[KeyValue::new("color", "green"), KeyValue::new("name", "lime")]);
+counter.add(5, &[KeyValue::new("color", "yellow"), KeyValue::new("name", "lemon")]);
+counter.add(8, &[KeyValue::new("name", "lemon"), KeyValue::new("color", "yellow")]); // bad perf
+```
+
+:heavy_check_mark: You should use consistent ordering of attributes to achieve
+the best performance.
+
+:heavy_check_mark: If feasible, provide the attributes in the sorted order of
```

Comment:
Rephrasing suggestion:
"If feasible, provide the attributes sorted by `Key`s in an ascending order to minimize the memory usage within Metrics SDK."